File: src/WellTool.Extra/hutool-extra/src/main/java/cn/hutool/extra/ssh/JschSessionPool.java
package cn.hutool.extra.ssh;

import cn.hutool.core.lang.SimpleCache;
import cn.hutool.core.map.SafeConcurrentHashMap;
import cn.hutool.core.util.StrUtil;
import com.jcraft.jsch.Session;

import java.util.Map;
import java.util.Map.Entry;

/**
 * Jsch会话池
 *
 * @author looly
 */
public enum JschSessionPool {

	/**
	 * 单例对象
	 */
	INSTANCE;

	/**
	 * SSH会话池，key：host，value：Session对象
	 */
	private final SimpleCache<String, Session> cache = new SimpleCache<>(new SafeConcurrentHashMap<>());

	/**
	 * 获取Session，不存在返回null
	 *
	 * @param key 键
	 * @return Session
	 */
	public Session get(String key) {
		return cache.get(key);
	}

	/**
	 * 获得一个SSH跳板机会话，重用已经使用的会话
	 *
	 * @param sshHost 跳板机主机
	 * @param sshPort 跳板机端口
	 * @param sshUser 跳板机用户名
	 * @param sshPass 跳板机密码
	 * @return SSH会话
	 */
	public Session getSession(String sshHost, int sshPort, String sshUser, String sshPass) {
		final String key = StrUtil.format("{}@{}:{}", sshUser, sshHost, sshPort);
		return this.cache.get(key, Session::isConnected, ()-> JschUtil.openSession(sshHost, sshPort, sshUser, sshPass));
	}

	/**
	 * 获得一个SSH跳板机会话，重用已经使用的会话
	 *
	 * @param sshHost    跳板机主机
	 * @param sshPort    跳板机端口
	 * @param sshUser    跳板机用户名
	 * @param prvkey     跳板机私钥路径
	 * @param passphrase 跳板机私钥密码
	 * @return SSH会话
	 */
	public Session getSession(String sshHost, int sshPort, String sshUser, String prvkey, byte[] passphrase) {
		final String key = StrUtil.format("{}@{}:{}", sshUser, sshHost, sshPort);
		return this.cache.get(key, Session::isConnected, ()->JschUtil.openSession(sshHost, sshPort, sshUser, prvkey, passphrase));
	}

	/**
	 * 获得一个SSH跳板机会话，重用已经使用的会话
	 *
	 * @param sshHost    跳板机主机
	 * @param sshPort    跳板机端口
	 * @param sshUser    跳板机用户名
	 * @param prvkey     跳板机私钥内容
	 * @param passphrase 跳板机私钥密码
	 * @return SSH会话
	 * @since 5.8.18
	 */
	public Session getSession(String sshHost, int sshPort, String sshUser, byte[] prvkey, byte[] passphrase) {
		final String key = StrUtil.format("{}@{}:{}", sshUser, sshHost, sshPort);
		return this.cache.get(key, Session::isConnected, ()->JschUtil.openSession(sshHost, sshPort, sshUser, prvkey, passphrase));
	}

	/**
	 * 加入Session
	 *
	 * @param key     键
	 * @param session Session
	 */
	public void put(String key, Session session) {
		this.cache.put(key, session);
	}

	/**
	 * 关闭SSH连接会话
	 *
	 * @param key 主机，格式为user@host:port
	 */
	public void close(String key) {
		Session session = get(key);
		if (session != null && session.isConnected()) {
			session.disconnect();
		}
		this.cache.remove(key);
	}

	/**
	 * 移除指定Session
	 *
	 * @param session Session会话
	 * @since 4.1.15
	 */
	public void remove(Session session) {
		if (null != session) {
			String key = null;
			for (Map.Entry<String, Session> entry : cache) {
				if (session.equals(entry.getValue())) {
					key = entry.getKey();
					break;
				}
			}

			if(null != key){
				cache.remove(key);
			}
		}
	}

	/**
	 * 关闭所有SSH连接会话
	 */
	public void closeAll() {
		Session session;
		for (Entry<String, Session> entry : this.cache) {
			session = entry.getValue();
			if (session != null && session.isConnected()) {
				session.disconnect();
			}
		}
		cache.clear();
	}
}

File: src/WellTool.Extra/Ssh/JschSessionPool.cs
using System.Text;

namespace WellTool.Extra.Ssh;

/// <summary>
/// Jsch会话池，用于管理SSH会话的重用
/// </summary>
public class JschSessionPool
{
    /// <summary>
    /// 获取单例实例
    /// </summary>
    public static JschSessionPool Instance { get; } = new JschSessionPool();

    private readonly Dictionary<string, ISshSession> _sessionCache = new();
    private readonly object _lock = new();

    /// <summary>
    /// 构造函数
    /// </summary>
    public JschSessionPool() { }

    /// <summary>
    /// 获取会话
    /// </summary>
    /// <param name="host">主机</param>
    /// <param name="port">端口</param>
    /// <param name="user">用户名</param>
    /// <param name="password">密码</param>
    /// <returns>SSH会话</returns>
    public ISshSession GetSession(string host, int port, string user, string password)
    {
        var key = GetSessionKey(host, port, user);
        lock (_lock)
        {
            if (_sessionCache.TryGetValue(key, out var session) && session.IsConnected)
            {
                return session;
            }

            var newSession = JschUtil.GetSession(host, port, user, password);
            newSession.Connect();
            _sessionCache[key] = newSession;
            return newSession;
        }
    }

    /// <summary>
    /// 获取会话(私钥认证)
    /// </summary>
    /// <param name="host">主机</param>
    /// <param name="port">端口</param>
    /// <param name="user">用户名</param>
    /// <param name="privateKeyPath">私钥路径</param>
    /// <param name="passphrase">私钥密码</param>
    /// <returns>SSH会话</returns>
    public ISshSession GetSession(string host, int port, string user, string privateKeyPath, string? passphrase)
    {
        var key = GetSessionKey(host, port, user);
        lock (_lock)
        {
            if (_sessionCache.TryGetValue(key, out var session) && session.IsConnected)
            {
                return session;
            }

            var newSession = JschUtil.GetSession(host, port, user, privateKeyPath, passphrase);
            newSession.Connect();
            _sessionCache[key] = newSession;
            return newSession;
        }
    }

    /// <summary>
    /// 获取会话(私钥认证)
    /// </summary>
    /// <param name="host">主机</param>
    /// <param name="port">端口</param>
    /// <param name="user">用户名</param>
    /// <param name="privateKey">私钥内容</param>
    /// <param name="passphrase">私钥密码</param>
    /// <returns>SSH会话</returns>
    public ISshSession GetSession(string host, int port, string user, byte[] privateKey, byte[]? passphrase)
    {
        var key = GetSessionKey(host, port, user);
        lock (_lock)
        {
            if (_sessionCache.TryGetValue(key, out var session) && session.IsConnected)
            {
                return session;
            }

            var newSession = JschUtil.GetSession(host, port, user, privateKey, passphrase);
            newSession.Connect();
            _sessionCache[key] = newSession;
            return newSession;
        }
    }

    /// <summary>
    /// 移除会话
    /// </summary>
    /// <param name="host">主机</param>
    /// <param name="port">端口</param>
    /// <param name="user">用户名</param>
    public void RemoveSession(string host, int port, string user)
    {
        var key = GetSessionKey(host, port, user);
        lock (_lock)
        {
            if (_sessionCache.TryGetValue(key, out var session))
            {
                session.Dispose();
                _sessionCache.Remove(key);
            }
        }
    }

    /// <summary>
    /// 清空所有会话
    /// </summary>
    public void Clear()
    {
        lock (_lock)
        {
            foreach (var session in _sessionCache.Values)
            {
                session.Dispose();
            }
            _sessionCache.Clear();
        }
    }

    private static string GetSessionKey(string host, int port, string user)
    {
        return $"{host}:{port}:{user}";
    }
}
